wandb/sdk/launch/sweeps/scheduler_optuna.py

```diff
@@ -0,0 +1,352 @@
+        with new versions, etc.
+        """
+
+        validate_optuna_config(self._sweep_config)
+
+        distributions = {}
+        for param, extras in self._sweep_config["parameters"].items():
+            print(param, extras)
+            if values := extras.get("values"):  # categorical
+                distributions[param] = CategoricalDistribution(values)
+            elif value := extras.get("value"):
+                distributions[param] = CategoricalDistribution([value])
+            elif type(extras.get("min")) == float:
+                log = "log" in param
+                distributions[param] = FloatDistribution(
+                    extras.get("min"), extras.get("max"), log=log
+                )
+            elif type(extras.get("min")) == int:
+                log = "log" in param
+                distributions[param] = IntDistribution(
+                    extras.get("min"), extras.get("max"), log=log
+                )
+            else:
+                print(f"{LOG_PREFIX} Unknown parameter type, help! {param=}, {extras=}")
+
+        # Do we want to set the metric here? Note for the users, we access the metric
+        # when pruning, so we need to know what its called. defaults to "loss"
+        # epochs defaults to epoch, could also name step...
+
+        return distributions
+
+    def _load_db(self, sweep_name: str = "example-study"):
+        """
+        Create an optuna study with a sqlite backened for loose state management
+        """
+        # TODO(gst): add to validate function to confirm this exists, warn user
+        params = self._sweep_config.get("parameters", {})
+        epochs = params.get("epochs", {}).get("value") or 20  # could also use STEP here
+        pruner_args = params.get("pruner", {})
+
+        pruner = optuna.pruners.HyperbandPruner(
+            min_resource=pruner_args.get(
+                "min_resource", 1
+            ),  # TODO(gst): defaults are from tutorial
+            max_resource=epochs,
+            reduction_factor=pruner_args.get("reduction_factor", 3),
+        )
+
+        storage_name = f"sqlite:///{sweep_name}.db"
+        return optuna.create_study(
+            study_name=sweep_name,
+            storage=storage_name,
+            pruner=pruner,
+            load_if_exists=True,
+        )
+
+    def _start(self) -> None:
+        for worker_id in range(self._num_workers):
+            print(f"{LOG_PREFIX}Starting AgentHeartbeat worker {worker_id}\n")
+            agent_config = self._api.register_agent(
+                f"{socket.gethostname()}-{worker_id}",  # host
+                sweep_id=self._sweep_id,
+                project_name=self._project,
+                entity=self._entity,
+            )
+            self._workers[worker_id] = _Worker(
+                agent_config=agent_config,
+                agent_id=agent_config["id"],
+            )
+
+    def _heartbeat(self, worker_id: int) -> None:
```

Comment:
is this still heartbeating to the server? we shouldnt need any heartbeat anymore right?